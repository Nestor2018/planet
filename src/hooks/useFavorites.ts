import { useState, useEffect } from "react";
import AsyncStorage from "@react-native-async-storage/async-storage";

const FAVORITES_KEY = "favorite_planets8";

interface Planet {
  id: string;
  [key: string]: any;
}

const useFavorites = () => {
  /**
   * State to store the list of favorite planets.
   */
  const [favorites, setFavorites] = useState<Planet[]>([]);

  useEffect(() => {
    /**
     * Loads the list of favorite planets from AsyncStorage.
     */
    const loadFavorites = async () => {
      const storedFavorites = await AsyncStorage.getItem(FAVORITES_KEY);
      if (storedFavorites) {
        setFavorites(JSON.parse(storedFavorites));
      }
    };
    loadFavorites();
  }, [favorites]);

  /**
   * Toggles the favorite status of a planet.
   * If the planet is already a favorite, it will be removed from the list.
   * If the planet is not a favorite, it will be added to the list.
   *
   * @param {Planet} planet - The planet to toggle in the favorites list.
   */
  const toggleFavorite = async (planet: Planet) => {
    const isFavorite = favorites.some(
      (favorite: Planet) => favorite.id === planet.id,
    );

    const updatedFavorites = isFavorite
      ? favorites.filter((favorite: Planet) => favorite.id !== planet.id)
      : [...favorites, planet];

    setFavorites(updatedFavorites);
    await AsyncStorage.setItem(FAVORITES_KEY, JSON.stringify(updatedFavorites));
  };

  return { favorites, toggleFavorite };
};

export default useFavorites;
