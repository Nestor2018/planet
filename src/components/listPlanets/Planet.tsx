import React from "react";
import { View, Text, TouchableOpacity, StyleSheet } from "react-native";
import { Ionicons } from "@expo/vector-icons";

import { PlanetType } from "../../interfaces/planet.interface";
import useFavorites from "../../hooks/useFavorites";
import colors from "../../utils/colors";

interface Props {
  planet: PlanetType;
  isFavorite: boolean;
}

const Planet: React.FC<Props> = ({ planet, isFavorite }) => {
  const { toggleFavorite } = useFavorites();
  return (
    <View style={styles.container}>
      <Text style={styles.text}>{planet.englishName}</Text>
      <TouchableOpacity onPress={() => toggleFavorite(planet)}>
        <Ionicons
          name={isFavorite ? "heart" : "heart-outline"}
          size={24}
          color={isFavorite ? colors.picton : colors.white}
        />
      </TouchableOpacity>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    justifyContent: "space-between",
    alignItems: "center",
    padding: 10,
    backgroundColor: colors.charade,
    borderRadius: 8,
    marginVertical: 5,
  },
  text: {
    color: colors.white,
    fontSize: 16,
  },
});

export default Planet;
