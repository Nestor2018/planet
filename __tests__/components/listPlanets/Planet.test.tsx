import React from "react";
import { render, fireEvent } from "@testing-library/react-native";
import Planet from "../../../src/components/listPlanets/Planet";
import useFavorites from "../../../src/hooks/useFavorites";

jest.mock("../../../src/hooks/useFavorites", () => ({
  __esModule: true,
  default: jest.fn(() => ({
    toggleFavorite: jest.fn(),
  })),
}));

describe("Planet Component", () => {
  const mockPlanet = {
    englishName: "Earth",
    id: "earth",
  };

  beforeEach(() => {
    (useFavorites as jest.Mock).mockClear();
  });

  it("renders planet name and favorite icon", () => {
    const { getByText, getByTestId } = render(
      <Planet planet={mockPlanet} isFavorite={false} />,
    );

    expect(getByText("Earth")).toBeTruthy();

    const icon = getByTestId("icon-heart-outline");
    expect(icon).toBeTruthy();
    expect(icon.props.size).toBe(24);
    expect(icon.props.color).toBe("#fff");
  });

  it("shows filled heart icon when planet is favorite", () => {
    const { getByTestId } = render(
      <Planet planet={mockPlanet} isFavorite={true} />,
    );

    const icon = getByTestId("icon-heart");
    expect(icon).toBeTruthy();
    expect(icon.props.color).toBe("#3c6fc8");
  });

  it("calls toggleFavorite when favorite button is pressed", () => {
    const mockToggleFavorite = jest.fn();
    (useFavorites as jest.Mock).mockReturnValue({
      toggleFavorite: mockToggleFavorite,
    });

    const { getByTestId } = render(
      <Planet planet={mockPlanet} isFavorite={false} />,
    );

    const button = getByTestId("icon-heart-outline");
    fireEvent.press(button);

    expect(mockToggleFavorite).toHaveBeenCalledTimes(1);
    expect(mockToggleFavorite).toHaveBeenCalledWith(mockPlanet);
  });

  it("matches snapshot", () => {
    const tree = render(
      <Planet planet={mockPlanet} isFavorite={false} />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
