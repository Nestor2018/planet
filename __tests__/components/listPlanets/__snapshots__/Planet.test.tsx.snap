// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Planet Component matches snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#272c35",
      "borderRadius": 8,
      "flexDirection": "row",
      "justifyContent": "space-between",
      "marginVertical": 5,
      "padding": 10,
    }
  }
>
  <Text
    style={
      {
        "color": "#fff",
        "fontSize": 16,
      }
    }
  >
    Earth
  </Text>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <view
      color="#fff"
      name="heart-outline"
      size={24}
      testID="icon-heart-outline"
    />
  </View>
</View>
`;
